### PR TITLE
Add global_counter to Trainer as optional argument

### PR DESCRIPTION
**Pull request by wielandbrendel**
This PR allows to use tf.train.exponential_decay together with Trainer. It adds an optional argument to pass a global_counter to Trainer, and to hence initialise the global_counter before initializing Trainer. This is necessary if the parameter to be decayed is part of the loss function.

### CODE REVIEW

File: tflearn/helpers/trainer.py [hunk 1]
```diff
@@ -87,8 +87,12 @@ def __init__(self, train_ops, graph=None, clip_gradients=5.0,
...
 
             self.global_loss = None
-            self.global_step = tf.Variable(0., name='Global_Step',
-                                           trainable=False)
+            if global_step:
+                self.global_step = global_step
+            else:
+                self.global_step = tf.Variable(0., name='Global_Step',
+                                               trainable=False)
+            self.global_step = global_step
```
Comment (braddengross) [inline tflearn/helpers/trainer.py]: This line shouldn't be necessary as it would have already been set.

File: tflearn/helpers/trainer.py [hunk 2]
```diff
@@ -87,8 +87,12 @@ def __init__(self, train_ops, graph=None, clip_gradients=5.0,
             self.validate_trainop_names()
 
             self.global_loss = None
-            self.global_step = tf.Variable(0., name='Global_Step',
-                                           trainable=False)
+            if global_step:
+                self.global_step = global_step
+            else:
```
Comment (braddengross) [inline tflearn/helpers/trainer.py]: This can be simplified into just checking if global_step = none and if so we create the new variable then assign it to self at the end.

Comment (braddengross) [inline tflearn/helpers/trainer.py]: Something like self.global_step = tf.Var if global_step is None else global_step

Comment (wielandbrendel) [inline tflearn/helpers/trainer.py]: I thought about that but putting everything in one line might hurt readability: after all, the definition of tf.Variable(...) is already taking two lines.

File: tflearn/helpers/trainer.py (hunk 1 above)
Comment (wielandbrendel) [inline tflearn/helpers/trainer.py]: The last line is a stupid mistake - sorry!

File: tflearn/helpers/trainer.py (hunk 2 above)
Comment (braddengross) [inline tflearn/helpers/trainer.py]: Sure, that makes sense. I was thinking more like we don't need the entire if else. It can be something like:
```
if global_step is None:
    global_step = tf.Variable(0., name='Global_Step', trainable=False)
self.global_step = global_step
```

but your way is more explicit.

File: tflearn/helpers/trainer.py [hunk 3]
```diff
@@ -87,8 +87,11 @@ def __init__(self, train_ops, graph=None, clip_gradients=5.0,
             self.validate_trainop_names()
 
             self.global_loss = None
-            self.global_step = tf.Variable(0., name='Global_Step',
-                                           trainable=False)
+            if global_step:
```
Comment (aymericdamien) [inline tflearn/helpers/trainer.py]: Here you need to set:
```
if global_step is None:
```
because 'if `Tensor`' raises an exception in TensorFlow.

**Comment by aymericdamien** [comment]
That looks good @wielandbrendel! I just added a comment to use:
```
if global_step is None:
    global_step = tf.Variable(0., name='Global_Step', trainable=False)
self.global_step = global_step
```
as @braddengross suggested, because the other way will raise a TensorFlow error.
